modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java

```diff
@@ -791,6 +843,167 @@ public synchronized void releaseDistributedTransaction(UUID sessionId) {
         return MessagingEngine.getInstance().getNextNMessageIdsInDLC(dlcQueueName, startMessageId, messageLimit);
     }
 
+    /***
+     * Get message IDs in DLC starting from given startMessageId up to the given message count.
+     *
+     * @param dlcQueueName Queue name of the Dead Letter Channel
+     * @param sourceQueue last message ID returned from invoking this method.
+     * @return List<Long> of message IDs moved to the DLC from the sourceQueue.
+     * @throws AndesException if an error occurs while reading messages from the database.
+     */
+    public int rerouteAllMessagesInDeadLetterChannelForQueue(String dlcQueueName, String sourceQueue, String
+            targetQueue, int internalBatchSize, boolean restoreToOriginalQueue) throws AndesException {
+
+        List<Long> currentMessageIdList;
+        Long lastMessageId = 0L;
+        int movedMessageCount = 0;
+        currentMessageIdList = getNextNMessageIdsInDLCForQueue(sourceQueue, dlcQueueName, lastMessageId,
+                internalBatchSize);
+        while (currentMessageIdList.size() > 0) {
+            int movedMessageCountInThisBatch = moveMessagesFromDLCToNewDestination(currentMessageIdList, sourceQueue,
+                    targetQueue, restoreToOriginalQueue);
+
+            if (log.isDebugEnabled()) {
+                log.debug("Successfully restored messages from sourceQueue : " + sourceQueue + " to targetQueue : "
+                        + targetQueue + " movedMessageCountInThisBatch : " + movedMessageCountInThisBatch);
+            }
+
+            movedMessageCount = movedMessageCount + movedMessageCountInThisBatch;
+            lastMessageId = currentMessageIdList.get(currentMessageIdList.size() - 1);
+
+            currentMessageIdList = getNextNMessageIdsInDLCForQueue(sourceQueue, dlcQueueName, lastMessageId,
+                    internalBatchSize);
+        }
+        return movedMessageCount;
+
+    }
+
+    /**
+     * Common method to restore a list of messages based on Id to its original queue or a different queue.
+     *
+     * @param messageIds             list of messages to be restored
+     * @param sourceQueue            original destination queue of the messages.
+     * @param targetQueue            new target destination of the messages.
+     * @param restoreToOriginalQueue true if the messages need to be restored to their original
+     *                               queues instead of a single target queue.
+     * @return int Number of messages that were successfully restored.
+     * @throws AndesException if the database calls to read/delete the messages/content fails.
+     */
+    public int moveMessagesFromDLCToNewDestination(List<Long> messageIds, String sourceQueue, String targetQueue,
+                                                    boolean restoreToOriginalQueue)
+            throws AndesException {
+
+        andesChannel = createChannel(new FlowControlListener() {
```

Comment:
Could we get rid of creating a channel to move the messages ? let's stick with moving them by updating the database